### PR TITLE
tests: Fix wsl test file mocks, set defaults for optional os-release fields

### PR DESCRIPTION
**Pull request by holmanb**
## Proposed Commit Message
```
tests: Fix wsl test

Force ds-identify to use the correct file mocks
Set defaults for optional os-release fields [1]

VERSION_ID and ID might not be set; set defaults.

[1] man 5 os-release

Fixes GH-5002
```

## Additional Context
Fixes GH-5002

## Test Steps
Comment out `VERSION_ID` in the host system's os-release file, run 

```bash
tox -e py3 -- tests/unittests/test_ds_identify.py::TestWSL
```

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Comment by holmanb** [comment]
Requesting review from @rvandegrift and @CarlosNihelton.

**Review by CarlosNihelton — APPROVED** [review]
LGTM 👍🏽

File: tools/ds-identify [hunk 1]
```diff
@@ -1654,8 +1654,8 @@ dscheck_WSL() {
     WSL_instance_name
     instance_name="${_RET}"
     # shellcheck source=/dev/null
-    . /etc/os-release
-    for userdatafile in "${instance_name}.user-data" "${ID}-${VERSION_ID}" "${ID}-all.user-data" "default.user-data"; do
+    . "${PATH_ROOT}/etc/os-release"
+    for userdatafile in "${instance_name}.user-data" "${ID:-linux}-${VERSION_ID:-}" "${ID:-linux}-all.user-data" "default.user-data"; do
```
Comment (blackboxsw) [inline tools/ds-identify:1658]: +1 and [TIL about ID=linux](https://www.freedesktop.org/software/systemd/man/latest/os-release.html#ID=) when unset

**Review by blackboxsw — APPROVED** [review]
Ship it

File: tools/ds-identify (hunk 1 above)
Comment (blackboxsw) [reply tools/ds-identify:1658]:
```suggestion
    for userdatafile in "${instance_name}.user-data" "${ID:-linux}-${VERSION_ID:-}.user-data" "${ID:-linux}-all.user-data" "default.user-data"; do
```

**Review by blackboxsw — CHANGES_REQUESTED** [review]
Found a bug here with previous code. ds-identify should have been matching "${ID}-${VERSION_ID}.user-data"

File: tools/ds-identify (hunk 1 above)
Comment (holmanb) [reply tools/ds-identify]: > It should be matching .user-data files throughout.

Nice catch!

Comment (blackboxsw) [reply tools/ds-identify]: I think the existing unittests were not specific enough in catching this failure path above as they touched all user-data files prior to tests and only exclusively tested the last one.
Here's a diff that'll better cover the different user-data files individually:
```diff
diff --git a/cloudinit/sources/DataSourceWSL.py b/cloudinit/sources/DataSourceWSL.py
index 8af6af244..0fccf61f7 100644
--- a/cloudinit/sources/DataSourceWSL.py
+++ b/cloudinit/sources/DataSourceWSL.py
@@ -131,13 +131,14 @@ def candidate_user_data_file_names(instance_name) -> List[str]:
     Return a list of candidate file names that may contain user-data
     in some supported format, ordered by precedence.
     """
-    distribution_id, version_id, _ = util.get_linux_distro()
+    distribution_id, version_id, version_codename = util.get_linux_distro()
+    version = version_id if version_id else version_codename
 
     return [
         # WSL instance specific:
         "%s.user-data" % instance_name,
         # release codename specific
-        "%s-%s.user-data" % (distribution_id, version_id),
+        "%s-%s.user-data" % (distribution_id, version),
         # distribution specific (Alpine, Arch, Fedora, openSUSE, Ubuntu...)
         "%s-all.user-data" % distribution_id,
         # generic, valid for all WSL distros and instances.
diff --git a/doc/rtd/reference/datasources/wsl.rst b/doc/rtd/reference/datasources/wsl.rst
index f0a40e19b..dbbfd500d 100644
--- a/doc/rtd/reference/datasources/wsl.rst
+++ b/doc/rtd/reference/datasources/wsl.rst
@@ -58,8 +58,10 @@ order specified below:
    instance named ``Sid-MLKit``.
 
 2. ``%USERPROFILE%\.cloud-init\<ID>-<VERSION_ID>.user-data`` for the
-   distro-specific configuration, matched by the distro ID and VERSION_CODENAME
-   entries as specified in ``/etc/os-release``. Example:
+   distro-specific configuration, matched by the distro ID and VERSION_ID
+   entries as specified in ``/etc/os-release``.  If VERSION_ID is not present,
+   then VERSION_CODENAME will be used instead.
+   Example:
    ``ubuntu-22.04.user-data`` will affect any instance created from an Ubuntu
    22.04 Jammy Jellyfish image if a more specific configuration file does not
    match.
diff --git a/tests/unittests/sources/test_wsl.py b/tests/unittests/sources/test_wsl.py
index a366ef713..9f5c40fcd 100644
--- a/tests/unittests/sources/test_wsl.py
+++ b/tests/unittests/sources/test_wsl.py
@@ -49,6 +49,7 @@ GOOD_MOUNTS = {
     },
 }
 SAMPLE_LINUX_DISTRO = ("ubuntu", "24.04", "noble")
+SAMPLE_LINUX_DISTRO_NO_VERSION_ID = ("debian", "", "trixie")
 
 
 class TestWSLHelperFunctions:
@@ -122,19 +123,37 @@ class TestWSLHelperFunctions:
         with pytest.raises(IOError):
             wsl.cmd_executable()
 
+    @pytest.mark.parametrize(
+        "linux_distro_value,files",
+        (
+            (
+                SAMPLE_LINUX_DISTRO,
+                [
+                    f"{INSTANCE_NAME}.user-data",
+                    "ubuntu-24.04.user-data",
+                    "ubuntu-all.user-data",
+                    "default.user-data",
+                ],
+            ),
+            (
+                SAMPLE_LINUX_DISTRO_NO_VERSION_ID,
+                [
+                    f"{INSTANCE_NAME}.user-data",
+                    "debian-trixie.user-data",
+                    "debian-all.user-data",
+                    "default.user-data",
+                ],
+            ),
+        ),
+    )
     @mock.patch("cloudinit.util.get_linux_distro")
-    def test_candidate_files(self, m_gld):
+    def test_candidate_files(self, m_gld, linux_distro_value, files):
         """
         Validate the file names candidate for holding user-data and their
         order of precedence.
         """
-        m_gld.return_value = SAMPLE_LINUX_DISTRO
-        assert [
-            f"{INSTANCE_NAME}.user-data",
-            "ubuntu-24.04.user-data",
-            "ubuntu-all.user-data",
-            "default.user-data",
-        ] == wsl.candidate_user_data_file_names(INSTANCE_NAME)
+        m_gld.return_value = linux_distro_value
+        assert files == wsl.candidate_user_data_file_names(INSTANCE_NAME)
 
     @pytest.mark.parametrize(
         "md_content,raises,errors,warnings,md_expected",
diff --git a/tests/unittests/test_ds_identify.py b/tests/unittests/test_ds_identify.py
index 56574d44d..1be00e464 100644
--- a/tests/unittests/test_ds_identify.py
+++ b/tests/unittests/test_ds_identify.py
@@ -131,6 +131,14 @@ MOCK_WSL_INSTANCE_DATA = {
         LOGO=ubuntu-logo
         """
     ),
+    "os_release_no_version_id": dedent(
+        """\
+        PRETTY_NAME="Debian GNU/Linux trixie/sid"
+        NAME="Debian GNU/Linux"
+        VERSION_CODENAME="trixie"
+        ID=debian
+        """
+    ),
 }
 
 shell_true = 0
@@ -1128,9 +1136,26 @@ class TestWSL(DsIdentifyBase):
         )
         return self._check_via_dict(data, RC_NOT_FOUND)
 
+    def test_found_via_userdata_version_codename(self):
+        """WLS datasource detected by VERSION_CODENAME when no VERSION_ID"""
+        data = copy.deepcopy(VALID_CFG["WSL-supported-debian"])
+        cloudinitdir = self.tmp_dir()
+        data["mocks"].append(
+            {
+                "name": "WSL_cloudinit_dir_in",
+                "ret": 0,
+                "RET": cloudinitdir,
+            },
+        )
+        filename = os.path.join(cloudinitdir, "debian-trixie.user-data")
+        Path(filename).touch()
+        self._check_via_dict(data, RC_FOUND, dslist=[data.get("ds"), DS_NONE])
+        Path(filename).unlink()
+
     def test_found_via_userdata(self):
-        """Asserts that WSL datasource is found if there is applicable
-        userdata files inside cloudinitdir."""
+        """
+        WSL datasource is found on applicable userdata files in cloudinitdir.
+        """
         data = copy.deepcopy(VALID_CFG["WSL-supported"])
         cloudinitdir = self.tmp_dir()
         data["mocks"].append(
@@ -1159,11 +1184,8 @@ class TestWSL(DsIdentifyBase):
             os.path.join(cloudinitdir, "default.user-data"),
         ]
 
-        # Ensures all applicable user data files exist.
         for filename in userdata_files:
             Path(filename).touch()
-
-        for filename in userdata_files:
             self._check_via_dict(
                 data, RC_FOUND, dslist=[data.get("ds"), DS_NONE]
             )
@@ -2236,4 +2258,29 @@ VALID_CFG = {
             "etc/os-release": MOCK_WSL_INSTANCE_DATA["os_release"],
         },
     },
+    "WSL-supported-debian": {
+        "ds": "WSL",
+        "mocks": [
+            MOCK_VIRT_IS_WSL,
+            MOCK_UNAME_IS_WSL,
+            {
+                "name": "WSL_instance_name",
+                "ret": 0,
+                "RET": MOCK_WSL_INSTANCE_DATA["name"],
+            },
+        ],
+        "files": {
+            "proc/mounts": (
+                "/dev/sdd / ext4 rw,errors=remount-ro,data=ordered 0 0\n"
+                "cgroup2 /sys/fs/cgroup cgroup2 rw,nosuid,nodev,noexec0 0\n"
+                "C:\\134 /mnt/c 9p rw,dirsync,aname=drvfs;path=C:\\;uid=0;"
+                "gid=0;symlinkroot=/mnt/...\n"
+                "snapfuse /snap/core22/1033 fuse.snapfuse ro,nodev,user_id=0,"
+                "group_id=0,allow_other 0 0"
+            ),
+            "etc/os-release": MOCK_WSL_INSTANCE_DATA[
+                "os_release_no_version_id"
+            ],
+        },
+    },
 }
diff --git a/tools/ds-identify b/tools/ds-identify
index dc3dc1810..741742c7f 100755
--- a/tools/ds-identify
+++ b/tools/ds-identify
@@ -1655,7 +1655,7 @@ dscheck_WSL() {
     instance_name="${_RET}"
     # shellcheck source=/dev/null
     . "${PATH_ROOT}/etc/os-release"
-    for userdatafile in "${instance_name}.user-data" "${ID:-linux}-${VERSION_ID:-}" "${ID:-linux}-all.user-data" "default.user-data"; do
+    for userdatafile in "${instance_name}.user-data" "${ID:-linux}-${VERSION_ID:-${VERSION_CODENAME}}".user-data "${ID:-linux}-all.user-data" "default.user-data"; do
         candidate="$cloudinitdir/$userdatafile"
         if [ -f "$candidate" ]; then
             debug 1 "Found applicable user data file for this instance at: $candidate"
```

**Review by blackboxsw — COMMENTED** [review]
Thanks @holmanb I added a diff suggestion unit test changes and a fallback to VERSION_CODENAME to help with debian /etc/os-release on trixie. let me know what you think

**Comment by holmanb** [comment]
> I added a diff suggestion unit test changes and a fallback to VERSION_CODENAME to help with debian /etc/os-release on trixie. let me know what you think

Looks good, that works for me!

**Review by blackboxsw — APPROVED** [review]
LGTM!

**Comment by rvandegrift** [comment]
@holmanb thanks, works for me

**Comment by holmanb** [comment]
Thanks for the reviews @rvandegrift @blackboxsw @CarlosNihelton!

**Pull request merged**